File: 4_6/p051.py
x, y = map(int, input().split())
M = 1000000007
c = x + y
child = 1
parent = 1
for i in range(1, c+1):
    child *= (i % M)
    child %= M
for i in range(1, x+1):
    parent *= (i % M)
    parent %= M
for i in range(1, y+1):
    parent *= (i % M)
    parent %= M
ans = 1
p = parent
for i in range(30):
    if M-2 & 1 << i:
        ans *= p
        ans %= M
    p *= p
    p %= M
ans = (child * ans) % M
print(ans)

File: 4_6/p051.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

var scan = bufio.NewScanner(os.Stdin)

const (
	initBufSize = 10000
	maxBufSize = 10000000
	M = 1000000007
)

func init() {
	buf := make([]byte, initBufSize)
	scan.Buffer(buf, maxBufSize)
	scan.Split(bufio.ScanWords)
}

func main() {
	x := NumStdin()
	y := NumStdin()

	fmt.Println(ncr(x+y, y))
}

func NumStdin() int {
	scan.Scan()
	num, err := strconv.Atoi(scan.Text())
	if err != nil {
		panic(err)
	}
	return num
}

func ncr(x, y int) int {
	lim := max(x, y)
	fact := make([]int, lim+1)
	fact[0] = 1
	for i := 1; i <= lim; i++ {
		fact[i] = (fact[i-1]*i)%M
	}
	return (fact[x] * pow((fact[x-y]*fact[y])%M, M-2)) % M
}

func max(a, b int) int {
	if a < b{
		return b
	}
	return a
}

func pow(a, b int) int {
	p := a
	ans := 1

	for i := 0; i <= 30; i++ {
		if b & (1 << i) != 0 {
			ans *= p
			ans %= M
		}
		p *= p
		p %= M
	}

	return ans
}
